

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Feature Extraction &mdash; protlearn  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Dimensionality Reduction &amp; Feature Selection" href="dimensionality_reduction.html" />
    <link rel="prev" title="Preprocessing" href="preprocessing.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Feature Extraction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#length">length</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parameters">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#returns">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#aac">aac</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#aaindex1">aaindex1</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id5">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7">Examples</a></li>
<li class="toctree-l3"><a class="reference internal" href="#notes">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ngram">ngram</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id8">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id9">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#entropy">entropy</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id11">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id13">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#apaac">apaac</a></li>
<li class="toctree-l2"><a class="reference internal" href="#atc">atc</a></li>
<li class="toctree-l2"><a class="reference internal" href="#binary">binary</a></li>
<li class="toctree-l2"><a class="reference internal" href="#csksaap">csksaap</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ctd">ctd</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ctdc">ctdc</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ctdd">ctdd</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ctdt">ctdt</a></li>
<li class="toctree-l2"><a class="reference internal" href="#geary">geary</a></li>
<li class="toctree-l2"><a class="reference internal" href="#moran">moran</a></li>
<li class="toctree-l2"><a class="reference internal" href="#moreau-broto">moreau_broto</a></li>
<li class="toctree-l2"><a class="reference internal" href="#motif">motif</a></li>
<li class="toctree-l2"><a class="reference internal" href="#paac">paac</a></li>
<li class="toctree-l2"><a class="reference internal" href="#posrich">posrich</a></li>
<li class="toctree-l2"><a class="reference internal" href="#qso">qso</a></li>
<li class="toctree-l2"><a class="reference internal" href="#socn">socn</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dimensionality_reduction.html">Dimensionality Reduction &amp; Feature Selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">protlearn</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Feature Extraction</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/feature_extraction.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="feature-extraction">
<span id="id1"></span><h1>Feature Extraction<a class="headerlink" href="#feature-extraction" title="Permalink to this headline">¶</a></h1>
<div class="section" id="length">
<h2>length<a class="headerlink" href="#length" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.length(X, *, method=&#39;int&#39;)
</pre></div>
</div>
<p>Sequence length in amino acids.</p>
<p>The number of amino acids that a protein or peptide is comprised of will be
counted and returned as either an integer (single sequence), array of
integers, or a one-hot-encoded array.</p>
<div class="section" id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>method: string, default=’int’</dt><dd><p>‘int’ : interval data <br />
‘ohe’ : one-hot encoded data</p>
</dd>
</dl>
</div>
<div class="section" id="returns">
<h3>Returns<a class="headerlink" href="#returns" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: int or ndarray of shape (n_samples, ) or (n_samples, n_unique_lengths)</dt><dd><p>Array containing sequence lengths.</p>
</dd>
</dl>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">length</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">,</span> <span class="s1">&#39;LLYPGP&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">l_int</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">l_int</span>
<span class="go">array([5, 7, 6])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">l_ohe</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ohe&#39;</span><span class="p">)</span>
<span class="go">array([[1., 0., 0.],</span>
<span class="go">       [0., 0., 1.],</span>
<span class="go">       [0., 1., 0.]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="aac">
<h2>aac<a class="headerlink" href="#aac" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.aac(X, *, method=&#39;relative&#39;, start=1, end=None)
</pre></div>
</div>
<p>Amino acid composition.</p>
<p>This function returns the frequency of amino acids for each sequence in the dataset.</p>
<p><cite>aac</cite> is calculated as follows:</p>
<div class="math notranslate nohighlight">
\[aac(i) = \frac{ n_i }{N}\]</div>
<p>where <em>i</em> denotes the 20 amino acid residues, <em>n</em><sub>i</sub> is the frequency of each
residue i, and <em>N</em> is the total number of residues in the sequence.</p>
<div class="section" id="id2">
<h3>Parameters<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>method: string, default=’relative’</dt><dd><p>‘absolute’ : absolute amino acid composition <br />
‘relative’ : relative amino acid composition</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start, this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id3">
<h3>Returns<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, n_unique_amino_acids)</dt><dd><p>Array containing the amino acid composition.</p>
</dd>
<dt>amino_acids: string</dt><dd><p>Corresponds to the columns (amino acids) in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id4">
<h3>Examples<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">aac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">comp</span><span class="p">,</span> <span class="n">aa</span> <span class="o">=</span> <span class="n">aac</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">comp</span>
<span class="go">array([[0.2       , 0.        , 0.        , 0.2       , 0.2       ,</span>
<span class="go">        0.        , 0.2       , 0.2       ],</span>
<span class="go">       [0.        , 0.28571429, 0.14285714, 0.14285714, 0.14285714,</span>
<span class="go">        0.14285714, 0.14285714, 0.        ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aa</span>
<span class="go">&#39;AEGKLPRY&#39;</span>
</pre></div>
</div>
<p>Note that columns containing all zeros have been removed from the final array.</p>
</div>
</div>
<div class="section" id="aaindex1">
<h2>aaindex1<a class="headerlink" href="#aaindex1" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.aaindex1(X, *, standardize=&#39;none&#39;, start=1, end=None)
</pre></div>
</div>
<p>AAIndex1-based physicochemical properties.</p>
<p>AAindex1 ver.9.2 (release Feb, 2017) is a set of 20 numerical values
representing various physicochemical and biological properties of amino
acids. Currently, it contains 566 indices, of which 553 contain no NaNs.
The indices will be collected for each amino acid in the sequence,
then averaged across the sequence.</p>
<p><cite>aaindex1</cite> is calculated as follows:</p>
<div class="math notranslate nohighlight">
\[aaindex1(i) = \frac{ \sum_{n=1}^{N}AAindex_i (aa_n) }{N}\]</div>
<p>where <em>i</em> denotes the 566 AAIndex1 indices, <em>aa</em><sub>n</sub> denotes the amino acid at
position <em>n</em>, and N is the total number of residues in the sequence.</p>
<div class="section" id="id5">
<h3>Parameters<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>standardize: string, default=’none’</dt><dd><p>‘none’ : unstandardized index matrix will be returned <br />
‘zscore’ : index matrix is standardized to have a mean of 0 and standard deviation of 1. <br />
‘minmax’ : index matrix is normalized to have a range of [0, 1].</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start, this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id6">
<h3>Returns<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, 553-566)</dt><dd><p>Array containing the AAIndex1 physicochemical properties.</p>
</dd>
<dt>desc: list of length 553-566</dt><dd><p>Corresponds to the columns (AAIndices) in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id7">
<h3>Examples<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">aaindex1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aaind</span><span class="p">,</span> <span class="n">inds</span> <span class="o">=</span> <span class="n">aaindex1</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">standardize</span><span class="o">=</span><span class="s1">&#39;zscore&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aaind</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 553)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">inds</span><span class="p">)</span>
<span class="go">553</span>
</pre></div>
</div>
</div>
<div class="section" id="notes">
<h3>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h3>
<p>Columns (indices) containing NaNs will be removed. Thus, the resulting index
matrix will have a column size between 553-566.</p>
</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Nakai, K., Kidera, A., and Kanehisa, M.; Cluster analysis of amino acid indices for prediction of protein structure and function. Protein Eng. 2, 93-100 (1988). [PMID:3244698]</p></li>
<li><p>Tomii, K. and Kanehisa, M.; Analysis of amino acid indices and mutation matrices for sequence comparison and structure prediction of proteins. Protein Eng. 9, 27-36 (1996). [PMID:9053899]</p></li>
<li><p>Kawashima, S., Ogata, H., and Kanehisa, M.; AAindex: amino acid index database. Nucleic Acids Res. 27, 368-369 (1999). [PMID:9847231]</p></li>
<li><p>Kawashima, S. and Kanehisa, M.; AAindex: amino acid index database. Nucleic Acids Res. 28, 374 (2000). [PMID:10592278]</p></li>
<li><p>Kawashima, S., Pokarowski, P., Pokarowska, M., Kolinski, A., Katayama, T., and Kanehisa, M.; AAindex: amino acid index database, progress report 2008. Nucleic Acids Res. 36, D202-D205 (2008). [PMID:17998252]</p></li>
</ul>
</div>
</div>
<div class="section" id="ngram">
<h2>ngram<a class="headerlink" href="#ngram" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.ngram(X, *, method=&#39;relative&#39;, start=1, end=None)
</pre></div>
</div>
<p>N-gram composition.</p>
<p>This function computes the di- or tripeptide composition of amino acid
sequences. Therefore, the function parameter <em>n</em> can only take on
the arguments 2 and 3 - otherwise, it will raise a ValueError.</p>
<div class="section" id="id8">
<h3>Parameters<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>n: int, default=2</dt><dd><p>Integer denoting the desired n-gram composition. <br />
2 : dipeptide composition <br />
3 : tripepitde composition</p>
</dd>
<dt>method: string, default=’relative’</dt><dd><p>‘absolute’: absolute n-gram composition <br />
‘relative’: relative n-gram composition</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id9">
<h3>Returns<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, n_unique^n)</dt><dd><p>Depending on n, the returned array will be of size: <br />
- (n_samples, 400) for dipeptide composition <br />
- (n_samples, 8000) for tripeptide composition <br />
if all possible n-gram combinations are represented.</p>
</dd>
<dt>n-grams: list of length n_unique^n</dt><dd><p>List of n-grams corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id10">
<h3>Examples<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">ngram</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">di</span><span class="p">,</span> <span class="n">ngrams</span> <span class="o">=</span> <span class="n">ngram</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">di</span>
<span class="go">array([[0.25      , 0.25      , 0.25      , 0.25      , 0.        ,</span>
<span class="go">        0.        , 0.        , 0.        , 0.        ],</span>
<span class="go">       [0.        , 0.        , 0.        , 0.16666667, 0.16666667,</span>
<span class="go">        0.16666667, 0.16666667, 0.16666667, 0.16666667]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ngrams</span>
<span class="go">[&#39;AR&#39;, &#39;KL&#39;, &#39;LY&#39;, &#39;RK&#39;, &#39;EE&#39;, &#39;ER&#39;, &#39;GL&#39;, &#39;KP&#39;, &#39;PG&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tri</span><span class="p">,</span> <span class="n">ngrams</span> <span class="o">=</span> <span class="n">ngram</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">array([[0.33333333, 0.33333333, 0.33333333, 0.        , 0.        ,</span>
<span class="go">        0.        , 0.        , 0.        ],</span>
<span class="go">       [0.        , 0.        , 0.        , 0.2       , 0.2       ,</span>
<span class="go">        0.2       , 0.2       , 0.2       ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ngrams</span>
<span class="go">[&#39;ARK&#39;, &#39;KLY&#39;, &#39;RKL&#39;, &#39;EER&#39;, &#39;ERK&#39;, &#39;KPG&#39;, &#39;PGL&#39;, &#39;RKP&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="entropy">
<h2>entropy<a class="headerlink" href="#entropy" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.entropy(X, *, standardize=&#39;none&#39;, start=1, end=None)
</pre></div>
</div>
<p>Shannon entropy.</p>
<p>This function computes the Shannon entropy for each sequence in the
dataset as follows:</p>
<div class="math notranslate nohighlight">
\[H(X) = -\sum_{i=1}^{20}P(x_i)log_2 P(x_i)\]</div>
<p>where <em>i</em> denotes the 20 amino acids and <em>P(x</em><sub>i</sub>) denotes the
probability of a given amino acid in the sequence.</p>
<div class="section" id="id11">
<h3>Parameters<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>standardize: string, default=’none’</dt><dd><p>‘none’ : unstandardized index matrix will be returned <br />
‘zscore’ : index matrix is standardized to have a mean of 0 and standard deviation of 1. <br />
‘minmax’ : index matrix is normalized to have a range of [0, 1].</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id12">
<h3>Returns<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples,) if len(X) &gt; 1, otherwise float</dt><dd><p>Array containing Shannon entropy values for each sequence.</p>
</dd>
</dl>
</div>
<div class="section" id="id13">
<h3>Examples<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">entropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ent</span> <span class="o">=</span> <span class="n">entropy</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ent</span>
<span class="go">array([2.32192809, 2.52164064])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="apaac">
<h2>apaac<a class="headerlink" href="#apaac" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="atc">
<h2>atc<a class="headerlink" href="#atc" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="binary">
<h2>binary<a class="headerlink" href="#binary" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="csksaap">
<h2>csksaap<a class="headerlink" href="#csksaap" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="ctd">
<h2>ctd<a class="headerlink" href="#ctd" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="ctdc">
<h2>ctdc<a class="headerlink" href="#ctdc" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="ctdd">
<h2>ctdd<a class="headerlink" href="#ctdd" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="ctdt">
<h2>ctdt<a class="headerlink" href="#ctdt" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="geary">
<h2>geary<a class="headerlink" href="#geary" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="moran">
<h2>moran<a class="headerlink" href="#moran" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="moreau-broto">
<h2>moreau_broto<a class="headerlink" href="#moreau-broto" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="motif">
<h2>motif<a class="headerlink" href="#motif" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="paac">
<h2>paac<a class="headerlink" href="#paac" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="posrich">
<h2>posrich<a class="headerlink" href="#posrich" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="qso">
<h2>qso<a class="headerlink" href="#qso" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
<div class="section" id="socn">
<h2>socn<a class="headerlink" href="#socn" title="Permalink to this headline">¶</a></h2>
<p>Coming soon!</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="dimensionality_reduction.html" class="btn btn-neutral float-right" title="Dimensionality Reduction &amp; Feature Selection" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="preprocessing.html" class="btn btn-neutral float-left" title="Preprocessing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Thomas Dorfer

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>