

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Feature Extraction &mdash; protlearn  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Dimensionality Reduction &amp; Feature Selection" href="dimensionality_reduction.html" />
    <link rel="prev" title="Preprocessing" href="preprocessing.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Feature Extraction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#length">length</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parameters">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#returns">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#aac">aac</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#aaindex1">aaindex1</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id5">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#notes">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id7">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ngram">ngram</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id8">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id9">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#entropy">entropy</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id11">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id13">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#posrich">posrich</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id14">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id15">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id16">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#motif">motif</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id18">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id20">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id21">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#atc">atc</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id22">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id25">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id26">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#binary">binary</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id27">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id28">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id29">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id30">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cksaap">cksaap</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id32">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id33">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id34">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id35">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id36">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ctd">ctd</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id37">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id38">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id39">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id40">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id41">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ctdc">ctdc</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id42">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id43">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id44">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id45">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ctdt">ctdt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id46">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id47">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id48">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id49">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ctdd">ctdd</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id50">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id51">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id52">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id53">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#geary">geary</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id54">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id55">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id56">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id57">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#moran">moran</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id58">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id59">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id60">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id61">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#moreau-broto">moreau_broto</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id62">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id63">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id64">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id65">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#paac">paac</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id66">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id67">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id68">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id69">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id70">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#apaac">apaac</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id71">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id72">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id73">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id74">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id75">Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="#socn">socn</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id76">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id77">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id78">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id79">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#qso">qso</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id80">Parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id81">Returns</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id82">Notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id83">References</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id84">Examples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dimensionality_reduction.html">Dimensionality Reduction &amp; Feature Selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">protlearn</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Feature Extraction</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/feature_extraction.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="feature-extraction">
<span id="id1"></span><h1>Feature Extraction<a class="headerlink" href="#feature-extraction" title="Permalink to this headline">¶</a></h1>
<div class="section" id="length">
<h2>length<a class="headerlink" href="#length" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.length(X, *, method=&#39;int&#39;)
</pre></div>
</div>
<p>Sequence length in amino acids.</p>
<p>The number of amino acids that a protein or peptide is comprised of will be
counted and returned as either an integer (single sequence), array of
integers, or a one-hot-encoded array.</p>
<div class="section" id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>method: string, default=’int’</dt><dd><p>‘int’ : interval data <br />
‘ohe’ : one-hot encoded data</p>
</dd>
</dl>
</div>
<div class="section" id="returns">
<h3>Returns<a class="headerlink" href="#returns" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: int or ndarray of shape (n_samples, ) or (n_samples, n_unique_lengths)</dt><dd><p>Array containing sequence lengths.</p>
</dd>
</dl>
</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">length</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">,</span> <span class="s1">&#39;LLYPGP&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">l_int</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">l_int</span>
<span class="go">array([5, 7, 6])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">l_ohe</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;ohe&#39;</span><span class="p">)</span>
<span class="go">array([[1., 0., 0.],</span>
<span class="go">       [0., 0., 1.],</span>
<span class="go">       [0., 1., 0.]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="aac">
<h2>aac<a class="headerlink" href="#aac" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.aac(X, *, method=&#39;relative&#39;, start=1, end=None)
</pre></div>
</div>
<p>Amino acid composition.</p>
<p>This function returns the frequency of amino acids for each sequence in the dataset.</p>
<p><cite>aac</cite> is calculated as follows:</p>
<div class="math notranslate nohighlight">
\[aac(i) = \frac{ n_i }{N}\]</div>
<p>where <em>i</em> denotes the 20 amino acid residues, <em>n</em><sub>i</sub> is the frequency of each
residue i, and <em>N</em> is the total number of residues in the sequence.</p>
<div class="section" id="id2">
<h3>Parameters<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>method: string, default=’relative’</dt><dd><p>‘absolute’ : absolute amino acid composition <br />
‘relative’ : relative amino acid composition</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start, this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id3">
<h3>Returns<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, n_unique_amino_acids)</dt><dd><p>Array containing the amino acid composition.</p>
</dd>
<dt>amino_acids: string</dt><dd><p>Corresponds to the columns (amino acids) in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id4">
<h3>Examples<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">aac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">comp</span><span class="p">,</span> <span class="n">aa</span> <span class="o">=</span> <span class="n">aac</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">comp</span>
<span class="go">array([[0.2       , 0.        , 0.        , 0.2       , 0.2       ,</span>
<span class="go">        0.        , 0.2       , 0.2       ],</span>
<span class="go">       [0.        , 0.28571429, 0.14285714, 0.14285714, 0.14285714,</span>
<span class="go">        0.14285714, 0.14285714, 0.        ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aa</span>
<span class="go">&#39;AEGKLPRY&#39;</span>
</pre></div>
</div>
<p>Note that columns containing all zeros have been removed from the final array.</p>
</div>
</div>
<div class="section" id="aaindex1">
<h2>aaindex1<a class="headerlink" href="#aaindex1" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.aaindex1(X, *, standardize=&#39;none&#39;, start=1, end=None)
</pre></div>
</div>
<p>AAIndex1-based physicochemical properties.</p>
<p>AAindex1 ver.9.2 (release Feb, 2017) is a set of 20 numerical values
representing various physicochemical and biological properties of amino
acids. Currently, it contains 566 indices, of which 553 contain no NaNs.
The indices will be collected for each amino acid in the sequence,
then averaged across the sequence.</p>
<p><cite>aaindex1</cite> is calculated as follows:</p>
<div class="math notranslate nohighlight">
\[aaindex1(i) = \frac{ \sum_{n=1}^{N}AAindex_i (aa_n) }{N}\]</div>
<p>where <em>i</em> denotes the 566 AAIndex1 indices, <em>aa</em><sub>n</sub> denotes the amino acid at
position <em>n</em>, and N is the total number of residues in the sequence.</p>
<div class="section" id="id5">
<h3>Parameters<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>standardize: string, default=’none’</dt><dd><p>‘none’ : unstandardized index matrix will be returned <br />
‘zscore’ : index matrix is standardized to have a mean of 0 and standard deviation of 1. <br />
‘minmax’ : index matrix is normalized to have a range of [0, 1].</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start, this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id6">
<h3>Returns<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, 553-566)</dt><dd><p>Array containing the AAIndex1 physicochemical properties.</p>
</dd>
<dt>desc: list of length 553-566</dt><dd><p>Corresponds to the columns (AAIndices) in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="notes">
<h3>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h3>
<p>Columns (indices) containing NaNs will be removed. Thus, the resulting index
matrix will have a column size between 553-566.</p>
</div>
<div class="section" id="references">
<h3>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Nakai, K., Kidera, A., and Kanehisa, M.; Cluster analysis of amino acid indices for prediction of protein structure and function. Protein Eng. 2, 93-100 (1988). [PMID:3244698]</p></li>
<li><p>Tomii, K. and Kanehisa, M.; Analysis of amino acid indices and mutation matrices for sequence comparison and structure prediction of proteins. Protein Eng. 9, 27-36 (1996). [PMID:9053899]</p></li>
<li><p>Kawashima, S., Ogata, H., and Kanehisa, M.; AAindex: amino acid index database. Nucleic Acids Res. 27, 368-369 (1999). [PMID:9847231]</p></li>
<li><p>Kawashima, S. and Kanehisa, M.; AAindex: amino acid index database. Nucleic Acids Res. 28, 374 (2000). [PMID:10592278]</p></li>
<li><p>Kawashima, S., Pokarowski, P., Pokarowska, M., Kolinski, A., Katayama, T., and Kanehisa, M.; AAindex: amino acid index database, progress report 2008. Nucleic Acids Res. 36, D202-D205 (2008). [PMID:17998252]</p></li>
</ul>
</div>
<div class="section" id="id7">
<h3>Examples<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">aaindex1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aaind</span><span class="p">,</span> <span class="n">inds</span> <span class="o">=</span> <span class="n">aaindex1</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">standardize</span><span class="o">=</span><span class="s1">&#39;zscore&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aaind</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 553)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">inds</span><span class="p">)</span>
<span class="go">553</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ngram">
<h2>ngram<a class="headerlink" href="#ngram" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.ngram(X, *, method=&#39;relative&#39;, start=1, end=None)
</pre></div>
</div>
<p>N-gram composition.</p>
<p>This function computes the di- or tripeptide composition of amino acid
sequences. Therefore, the function parameter <em>n</em> can only take on
the arguments 2 and 3 - otherwise, it will raise a ValueError.</p>
<div class="section" id="id8">
<h3>Parameters<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>n: int, default=2</dt><dd><p>Integer denoting the desired n-gram composition. <br />
2 : dipeptide composition <br />
3 : tripepitde composition</p>
</dd>
<dt>method: string, default=’relative’</dt><dd><p>‘absolute’: absolute n-gram composition <br />
‘relative’: relative n-gram composition</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id9">
<h3>Returns<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, n_unique^n)</dt><dd><p>Depending on n, the returned array will be of size: <br />
- (n_samples, 400) for dipeptide composition <br />
- (n_samples, 8000) for tripeptide composition <br />
if all possible n-gram combinations are represented.</p>
</dd>
<dt>n-grams: list of length n_unique^n</dt><dd><p>List of n-grams corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id10">
<h3>Examples<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">ngram</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">di</span><span class="p">,</span> <span class="n">ngrams</span> <span class="o">=</span> <span class="n">ngram</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">di</span>
<span class="go">array([[0.25      , 0.25      , 0.25      , 0.25      , 0.        ,</span>
<span class="go">        0.        , 0.        , 0.        , 0.        ],</span>
<span class="go">       [0.        , 0.        , 0.        , 0.16666667, 0.16666667,</span>
<span class="go">        0.16666667, 0.16666667, 0.16666667, 0.16666667]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ngrams</span>
<span class="go">[&#39;AR&#39;, &#39;KL&#39;, &#39;LY&#39;, &#39;RK&#39;, &#39;EE&#39;, &#39;ER&#39;, &#39;GL&#39;, &#39;KP&#39;, &#39;PG&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tri</span><span class="p">,</span> <span class="n">ngrams</span> <span class="o">=</span> <span class="n">ngram</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">array([[0.33333333, 0.33333333, 0.33333333, 0.        , 0.        ,</span>
<span class="go">        0.        , 0.        , 0.        ],</span>
<span class="go">       [0.        , 0.        , 0.        , 0.2       , 0.2       ,</span>
<span class="go">        0.2       , 0.2       , 0.2       ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ngrams</span>
<span class="go">[&#39;ARK&#39;, &#39;KLY&#39;, &#39;RKL&#39;, &#39;EER&#39;, &#39;ERK&#39;, &#39;KPG&#39;, &#39;PGL&#39;, &#39;RKP&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="entropy">
<h2>entropy<a class="headerlink" href="#entropy" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.entropy(X, *, standardize=&#39;none&#39;, start=1, end=None)
</pre></div>
</div>
<p>Shannon entropy.</p>
<p>This function computes the Shannon entropy for each sequence in the
dataset as follows:</p>
<div class="math notranslate nohighlight">
\[H(X) = -\sum_{i=1}^{20}P(x_i)log_2 P(x_i)\]</div>
<p>where <em>i</em> denotes the 20 amino acids and <em>P(x</em><sub>i</sub>) denotes the
probability of a given amino acid in the sequence.</p>
<div class="section" id="id11">
<h3>Parameters<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>standardize: string, default=’none’</dt><dd><p>‘none’ : unstandardized index matrix will be returned <br />
‘zscore’ : index matrix is standardized to have a mean of 0 and standard deviation of 1. <br />
‘minmax’ : index matrix is normalized to have a range of [0, 1].</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id12">
<h3>Returns<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples,) if len(X) &gt; 1, otherwise float</dt><dd><p>Array containing Shannon entropy values for each sequence.</p>
</dd>
</dl>
</div>
<div class="section" id="id13">
<h3>Examples<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">entropy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">,</span> <span class="s1">&#39;AAAAAALY&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ent</span> <span class="o">=</span> <span class="n">entropy</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ent</span>
<span class="go">array([2.32192809, 2.52164064, 0.64020643])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="posrich">
<h2>posrich<a class="headerlink" href="#posrich" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.posrich(X, *, position, aminoacid)
</pre></div>
</div>
<p>Position-specific amino acids.</p>
<p>This function returns a binary vector or matrix in which ones indicate the
presence of the given amino acid(s) at the specified position(s), and zeros
indicate their absence.</p>
<div class="section" id="id14">
<h3>Parameters<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>position: int or list</dt><dd><p>Integer or list of integers denoting the position(s) in the sequence.</p>
</dd>
<dt>aminoacid: string or list</dt><dd><p>String or list of strings indicating the amino acid(s) of interest.</p>
</dd>
</dl>
</div>
<div class="section" id="id15">
<h3>Returns<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, ) or (n_samples, n_positions)</dt><dd><p>Binary vector indicating position-specific presence of amino acids.</p>
</dd>
</dl>
</div>
<div class="section" id="id16">
<h3>Notes<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h3>
<p>The position argument is based on one-based indexing.</p>
</div>
<div class="section" id="id17">
<h3>Examples<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">posrich</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;ERNLAPG&#39;</span><span class="p">,</span> <span class="s1">&#39;YRLQLLLY&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos_single</span> <span class="o">=</span> <span class="n">posrich</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">aminoacid</span><span class="o">=</span><span class="s1">&#39;L&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos_single</span>
<span class="go">array([1., 1., 0.])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos_multiple</span> <span class="o">=</span> <span class="n">posrich</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="n">aminoacid</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;R&#39;</span><span class="p">,</span><span class="s1">&#39;N&#39;</span><span class="p">,</span><span class="s1">&#39;L&#39;</span><span class="p">])</span>
<span class="go">array([[1., 0., 1.],</span>
<span class="go">       [1., 1., 1.],</span>
<span class="go">       [1., 0., 0.]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="motif">
<h2>motif<a class="headerlink" href="#motif" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.motif(X, pattern, *, start=1, end=None)
</pre></div>
</div>
<p>Sequence motifs.</p>
<p>This function returns a binary vector indicating the presence of a specified
amino acid sequence motif.</p>
<div class="section" id="id18">
<h3>Parameters<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>pattern: string</dt><dd><p>Represents the sequence motif. <br />
x –&gt; any amino acid <br />
[XY] –&gt; X or Y <br />
{X} –&gt; any amino acid except X</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id19">
<h3>Returns<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples,)</dt><dd><p>Binary vector indicating the presence of the motif in sequences.</p>
</dd>
</dl>
</div>
<div class="section" id="id20">
<h3>Notes<a class="headerlink" href="#id20" title="Permalink to this headline">¶</a></h3>
<p>Based on the example above, ‘pattern1’ is interpreted as follows:
Two consecutive amino acids ‘A’, followed by any amino acid, followed by
either a ‘K’ or a ‘C’.</p>
<p>Likewise, pattern2 is interpreted as follows:
Any two consecutive amino acids, followed by a ‘C’, followed by either a ‘D’
or an ‘A’, followed by any two amino acids, followed by any amino acid
except ‘Y’.</p>
</div>
<div class="section" id="id21">
<h3>Examples<a class="headerlink" href="#id21" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">motif</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AARKYLL&#39;</span><span class="p">,</span> <span class="s1">&#39;LELCDPGPG&#39;</span><span class="p">,</span> <span class="s1">&#39;RAAANCDD&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pattern1</span> <span class="o">=</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s1">&#39;AAx[KC]&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m1</span> <span class="o">=</span> <span class="n">motif</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">pattern1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m1</span>
<span class="go">array([1., 0., 1.])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pattern2</span> <span class="o">=</span> <span class="s1">&#39;xxC[DA]xx</span><span class="si">{Y}</span><span class="s1">&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m2</span> <span class="o">=</span> <span class="n">motif</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">pattern2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m2</span>
<span class="go">array([0., 1., 0.])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="atc">
<h2>atc<a class="headerlink" href="#atc" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.atc(X, *, method=&#39;relative&#39;, start=1, end=None)
</pre></div>
</div>
<p>Atomic and bond composition.</p>
<p>This function returns the sum of atomic and bond compositions for each
amino acid sequence. The atomic features are comprised of five atoms
(C, H, N, O, and S), and the bond features are comprised of total bonds,
single bonds, and double bonds.</p>
<div class="section" id="id22">
<h3>Parameters<a class="headerlink" href="#id22" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>method: string, default=’relative’</dt><dd><p>‘absolute’: absolute atomic composition <br />
‘relative’: relative atomic composition</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id23">
<h3>Returns<a class="headerlink" href="#id23" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr_atoms:  ndarray of shape (n_samples, 5)</dt><dd><p>Array containing atomic compositions.</p>
</dd>
<dt>arr_bonds: ndarray of shape (n_samples, 3)</dt><dd><p>Array containing bond compositions.</p>
</dd>
</dl>
</div>
<div class="section" id="id24">
<h3>Notes<a class="headerlink" href="#id24" title="Permalink to this headline">¶</a></h3>
<p>The ‘method’ argument only applies to the atomic composition, not
the bond composition.</p>
</div>
<div class="section" id="id25">
<h3>References<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Kumar, R., Chaudhary, K., Singh Chauhan, J. et al. An in silico platform for predicting, screening and designing of antihypertensive peptides. Sci Rep 5, 12512 (2015). <a class="reference external" href="https://doi.org/10.1038/srep12512">https://doi.org/10.1038/srep12512</a></p></li>
</ul>
</div>
<div class="section" id="id26">
<h3>Examples<a class="headerlink" href="#id26" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">atc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">,</span> <span class="s1">&#39;AAAAAALY&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">atoms</span><span class="p">,</span> <span class="n">bonds</span> <span class="o">=</span> <span class="n">atc</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">atoms</span>
<span class="go">array([[0.27522936, 0.5412844 , 0.08256881, 0.10091743, 0.],</span>
<span class="go">       [0.25547445, 0.53284672, 0.08029197, 0.13138686, 0.],</span>
<span class="go">       [0.26612903, 0.53225806, 0.06451613, 0.13709677, 0.]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bonds</span>
<span class="go">array([[105.,  96.,   9.],</span>
<span class="go">       [131., 121.,  10.],</span>
<span class="go">       [117., 106.,  11.]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="binary">
<h2>binary<a class="headerlink" href="#binary" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.binary(X, *, padding=False, start=1, end=None)
</pre></div>
</div>
<p>Binary profile pattern.</p>
<p>This function returns the binary profile pattern for each amino acid
sequence in the dataset. The output array is therefore very sparse, as each
amino acid is represented by a 20-dimensional vector with only one non-zero
value.</p>
<div class="section" id="id27">
<h3>Parameters<a class="headerlink" href="#id27" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>padding: bool, default=False</dt><dd><p>Pad sequences of unequal lengths with zeros at the posterior end.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id28">
<h3>Returns<a class="headerlink" href="#id28" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, 20*seq_length)</dt><dd><p>Array containing binary profile pattern.</p>
</dd>
</dl>
</div>
<div class="section" id="id29">
<h3>Notes<a class="headerlink" href="#id29" title="Permalink to this headline">¶</a></h3>
<p>This function is intended for proteins or peptides with equal lengths only.</p>
</div>
<div class="section" id="id30">
<h3>References<a class="headerlink" href="#id30" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Ansari, H. R., &amp; Raghava, G. P. (2010). Identification of conformational B-cell Epitopes in an antigen from its primary sequence. Immunome research, 6, 6. <a class="reference external" href="https://doi.org/10.1186/1745-7580-6-6">https://doi.org/10.1186/1745-7580-6-6</a></p></li>
<li><p>Chen Z, Chen Y-Z, Wang X-F, Wang C, Yan R-X, Zhang Z (2011) Prediction of Ubiquitination Sites by Using the Composition of k-Spaced Amino Acid Pairs. PLoS ONE 6(7): e22930. <a class="reference external" href="https://doi.org/10.1371/journal.pone.0022930">https://doi.org/10.1371/journal.pone.0022930</a></p></li>
</ul>
</div>
<div class="section" id="id31">
<h3>Examples<a class="headerlink" href="#id31" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">binary</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bpp</span> <span class="o">=</span> <span class="n">binary</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">padding</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bpp</span>
<span class="go">array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],</span>
<span class="go">       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,</span>
<span class="go">        0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="cksaap">
<h2>cksaap<a class="headerlink" href="#cksaap" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.cksaap(X, *, k=1, start=1, end=None)
</pre></div>
</div>
<p>Composition of k-spaced amino acid pairs.</p>
<p>This function returns the k-spaced amino acid pair composition of each
sequence in the dataset. Since there are 20 natural amino acids, there are
400 possible amino acid pairs. The parameter ‘k’ represents the gap between
the amino acid pair. An example for k=1 would be AxY, where ‘x’ can be any
amino acid. Similary, an example for k=2 would be AxxY. If k=0, the function
returns the dipeptide composition of each sequence.</p>
<div class="section" id="id32">
<h3>Parameters<a class="headerlink" href="#id32" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt><a href="#id85"><span class="problematic" id="id86">lambda_</span></a>: int, default=1</dt><dd><p>Counted rank (tier) of the correlation along an amino acid sequence.</p>
</dd>
<dt>k: int, default=1</dt><dd><p>Space between two amino acid pairs.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id33">
<h3>Returns<a class="headerlink" href="#id33" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, 400)</dt><dd><p>Array containing k-spaced amino acid pair composition.</p>
</dd>
<dt>patterns: list of length 400</dt><dd><p>Amino acid pairs with k gaps corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id34">
<h3>Notes<a class="headerlink" href="#id34" title="Permalink to this headline">¶</a></h3>
<p>Columns containing only zeros will be deleted. Therefore, the returned
array and list may have a lower dimensionality than 400, depending on the
unique number of k-spaced amino acid pairs in the dataset.</p>
</div>
<div class="section" id="id35">
<h3>References<a class="headerlink" href="#id35" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Chen, K., Kurgan, L.A. &amp; Ruan, J. Prediction of flexible/rigid regions from protein sequences using k-spaced amino acid pairs. BMC Struct Biol 7, 25 (2007). <a class="reference external" href="https://doi.org/10.1186/1472-6807-7-25">https://doi.org/10.1186/1472-6807-7-25</a></p></li>
</ul>
</div>
<div class="section" id="id36">
<h3>Examples<a class="headerlink" href="#id36" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">cksaap</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">,</span> <span class="s1">&#39;AAAAAALY&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ck</span><span class="p">,</span> <span class="n">pairs</span> <span class="o">=</span> <span class="n">cksaap</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ck</span>
<span class="go">array([[0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0],</span>
<span class="go">       [0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1],</span>
<span class="go">       [4, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pairs</span>
<span class="go">[&#39;A.A&#39;, &#39;A.K&#39;, &#39;A.L&#39;, &#39;A.Y&#39;, &#39;E.K&#39;, &#39;E.R&#39;, &#39;K.G&#39;, &#39;K.Y&#39;, &#39;P.L&#39;, &#39;R.L&#39;, &#39;R.P&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ck2</span><span class="p">,</span> <span class="n">pairs2</span> <span class="o">=</span> <span class="n">cksaap</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ck2</span>
<span class="go">array([[0, 1, 0, 0, 0, 0, 0, 1],</span>
<span class="go">       [0, 0, 0, 1, 1, 1, 1, 0],</span>
<span class="go">       [3, 1, 1, 0, 0, 0, 0, 0]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pairs2</span>
<span class="go">[&#39;A..A&#39;, &#39;A..L&#39;, &#39;A..Y&#39;, &#39;E..K&#39;, &#39;E..P&#39;, &#39;K..L&#39;, &#39;R..G&#39;, &#39;R..Y&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ctd">
<h2>ctd<a class="headerlink" href="#ctd" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.ctd(X, *, start=1, end=None)
</pre></div>
</div>
<p>Conjoint triad descriptors.</p>
<p>These descriptors were initially developed to model protein-protein
interactions. Amino acids can be grouped into 7 different classes based on
their dipoles and side chain volumes, which reflect their electrostatic and
hydrophobic interactions. After grouping, these class triads are counted and
normalized (for more information, see <a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>).</p>
<div class="section" id="id37">
<h3>Parameters<a class="headerlink" href="#id37" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id38">
<h3>Returns<a class="headerlink" href="#id38" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, 343)</dt><dd><p>Array containing conjoint triad descriptors.</p>
</dd>
<dt>ctd_list: list of length 343</dt><dd><p>Unique class triads corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id39">
<h3>Notes<a class="headerlink" href="#id39" title="Permalink to this headline">¶</a></h3>
<p>Columns containing only zeros will be deleted. Therefore, the returned
array and list may have a lower dimensionality than 343, depending on the
unique number conjoint triad descriptors in the dataset.</p>
</div>
<div class="section" id="id40">
<h3>References<a class="headerlink" href="#id40" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Shen J, Zhang J, Luo X, Zhu W, Yu K, Chen K, Li Y, Jiang H (2007) Predicting protein-protein interactions based only on sequences information. Proc Natl Acad Sci USA 104: 4337 – 4341</p></li>
</ul>
</div>
<div class="section" id="id41">
<h3>Examples<a class="headerlink" href="#id41" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">ctd</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKKLYLYL&#39;</span><span class="p">,</span> <span class="s1">&#39;EEEERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ctd_arr</span><span class="p">,</span> <span class="n">ctd_desc</span> <span class="o">=</span> <span class="n">ctd</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ctd_arr</span>
<span class="go">array([[1., 2., 1., 1., 1., 1., 0., 0., 0., 0., 0.],</span>
<span class="go">        [0., 0., 0., 0., 1., 0., 1., 1., 1., 1., 2.]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ctd_desc</span>
<span class="go">[&#39;155&#39;, &#39;232&#39;, &#39;323&#39;, &#39;523&#39;, &#39;552&#39;, &#39;555&#39;, &#39;212&#39;, &#39;521&#39;, &#39;655&#39;, &#39;665&#39;, &#39;666&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ctdc">
<h2>ctdc<a class="headerlink" href="#ctdc" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.ctdc(X, *, start=1, end=None)
</pre></div>
</div>
<p>Composition/Transition/Distribution - Composition.</p>
<p>Amino acids are categorized into 3 groups based on their physicochemical
properties. The properties used here include hydrophobicity, normalized van
der Waals volume, polarity, polarizability, charge, secondary structure, and
solvent accessibility. For hydrophobicity, we use seven different groupings
based on different studies, which can all be found in AAIndex1.</p>
<p>After grouping, the frequency of each class will be calculated for each
physicochemical property per sequence in the dataset. For instance, the
sequence ‘ARKLY’ translates to ‘23311’ with respect to polarity groups.
Thus, for polarity, the outcome will be P1 = 2/5, P2 = 1/5, and P3 = 2/5.</p>
<p>As there are 13 different properties and 3 groups for each, the total dimension
of this descriptor is 39.</p>
<div class="section" id="id42">
<h3>Parameters<a class="headerlink" href="#id42" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id43">
<h3>Returns<a class="headerlink" href="#id43" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, 39)</dt><dd><p>Array containing grouped composition of physicochemical properties.</p>
</dd>
<dt>desc: list of length 39</dt><dd><p>Descriptor properties corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id44">
<h3>References<a class="headerlink" href="#id44" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Dubchak, I., Muchnik, I., Holbrook, S. R. &amp; Kim, S.-H. Prediction of protein folding class using global description of amino acid sequence. Proceedings of the National Academy of Sciences 92, 8700–8704 (1995).</p></li>
<li><p>Dubchak, I., Muchnik, I., Mayor, C., Dralyuk, I. &amp; Kim, S.-H. Recognition of a protein fold in the context of the scop classification. Proteins: Structure, Function, and Bioinformatics 35, 401–407 (1999).</p></li>
</ul>
</div>
<div class="section" id="id45">
<h3>Examples<a class="headerlink" href="#id45" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">ctdc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">ctdc</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span>
<span class="go">array([[0.        , 0.6       , 0.4       , 0.4       , 0.6       ,</span>
<span class="go">        0.        , 0.6       , 0.2       , 0.2       , 0.4       ,</span>
<span class="go">        0.        , 0.6       , 0.2       , 0.4       , 0.4       ,</span>
<span class="go">        0.4       , 0.4       , 0.2       , 0.4       , 0.2       ,</span>
<span class="go">        0.4       , 0.2       , 0.2       , 0.6       , 0.4       ,</span>
<span class="go">        0.2       , 0.4       , 0.2       , 0.2       , 0.6       ,</span>
<span class="go">        0.4       , 0.6       , 0.        , 0.8       , 0.2       ,</span>
<span class="go">        0.        , 0.4       , 0.4       , 0.2       ],</span>
<span class="go">       [0.42857143, 0.28571429, 0.28571429, 0.85714286, 0.14285714,</span>
<span class="go">        0.        , 0.71428571, 0.14285714, 0.14285714, 0.57142857,</span>
<span class="go">        0.28571429, 0.14285714, 0.57142857, 0.28571429, 0.14285714,</span>
<span class="go">        0.57142857, 0.28571429, 0.14285714, 0.57142857, 0.14285714,</span>
<span class="go">        0.28571429, 0.28571429, 0.42857143, 0.28571429, 0.14285714,</span>
<span class="go">        0.28571429, 0.57142857, 0.14285714, 0.57142857, 0.28571429,</span>
<span class="go">        0.28571429, 0.42857143, 0.28571429, 0.71428571, 0.        ,</span>
<span class="go">        0.28571429, 0.28571429, 0.57142857, 0.14285714]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span>
<span class="go">[&#39;Hydrophobicity_ARGP820101-G1&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ARGP820101-G2&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ARGP820101-G3&#39;,</span>
<span class="go"> &#39;Hydrophobicity_CASG920101-G1&#39;,</span>
<span class="go"> &#39;Hydrophobicity_CASG920101-G2&#39;,</span>
<span class="go"> &#39;Hydrophobicity_CASG920101-G3&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ENGD860101-G1&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ENGD860101-G2&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ENGD860101-G3&#39;,</span>
<span class="go"> &#39;Hydrophobicity_FASG890101-G1&#39;,</span>
<span class="go"> &#39;Hydrophobicity_FASG890101-G2&#39;,</span>
<span class="go"> &#39;Hydrophobicity_FASG890101-G3&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PONP930101-G1&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PONP930101-G2&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PONP930101-G3&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PRAM900101-G1&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PRAM900101-G2&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PRAM900101-G3&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ZIMJ680101-G1&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ZIMJ680101-G2&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ZIMJ680101-G3&#39;,</span>
<span class="go"> &#39;Normalized van der Waals Volume-G1&#39;,</span>
<span class="go"> &#39;Normalized van der Waals Volume-G2&#39;,</span>
<span class="go"> &#39;Normalized van der Waals Volume-G3&#39;,</span>
<span class="go"> &#39;Polarity-G1&#39;,</span>
<span class="go"> &#39;Polarity-G2&#39;,</span>
<span class="go"> &#39;Polarity-G3&#39;,</span>
<span class="go"> &#39;Polarizability-G1&#39;,</span>
<span class="go"> &#39;Polarizability-G2&#39;,</span>
<span class="go"> &#39;Polarizability-G3&#39;,</span>
<span class="go"> &#39;Charge-G1&#39;,</span>
<span class="go"> &#39;Charge-G2&#39;,</span>
<span class="go"> &#39;Charge-G3&#39;,</span>
<span class="go"> &#39;Secondary structure-G1&#39;,</span>
<span class="go"> &#39;Secondary structure-G2&#39;,</span>
<span class="go"> &#39;Secondary structure-G3&#39;,</span>
<span class="go"> &#39;Solvent accessibility-G1&#39;,</span>
<span class="go"> &#39;Solvent accessibility-G2&#39;,</span>
<span class="go"> &#39;Solvent accessibility-G3&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ctdt">
<h2>ctdt<a class="headerlink" href="#ctdt" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.ctdt(X, *, start=1, end=None)
</pre></div>
</div>
<p>Composition/Transition/Distribution - Transition</p>
<p>Amino acids are categorized into 3 groups based on their physicochemical
properties. The properties used here include hydrophobicity, normalized van
der Waals volume, polarity, polarizability, charge, secondary structure, and
solvent accessibility. For hydrophobicity, we use seven different groupings
based on different studies, which can all be found in AAIndex1.</p>
<p>This descriptor computes the frequency of transitions between groups. For
instance, if the encoded sequence is ‘32132223311311222222’, then there are
2 transitions from groups 1 to 2 and 2 to 1. Therefore, the descriptor for
this particular group transition will be 2/19. Similarly, there are 3
transitions from groups 2 to 3 and 3 to 2, so the descriptor for this
transition is 3/19.</p>
<p>As with CTDC, the dimensionality of this feature is 39.</p>
<div class="section" id="id46">
<h3>Parameters<a class="headerlink" href="#id46" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id47">
<h3>Returns<a class="headerlink" href="#id47" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, 39)</dt><dd><p>Array containing group transitions of physicochemical properties.</p>
</dd>
<dt>desc: list of length 39</dt><dd><p>Order of transition groups corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id48">
<h3>References<a class="headerlink" href="#id48" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Dubchak, I., Muchnik, I., Holbrook, S. R. &amp; Kim, S.-H. Prediction of protein folding class using global description of amino acid sequence. Proceedings of the National Academy of Sciences 92, 8700–8704 (1995).</p></li>
<li><p>Dubchak, I., Muchnik, I., Mayor, C., Dralyuk, I. &amp; Kim, S.-H. Recognition of a protein fold in the context of the scop classification. Proteins: Structure, Function, and Bioinformatics 35, 401–407 (1999).</p></li>
</ul>
</div>
<div class="section" id="id49">
<h3>Examples<a class="headerlink" href="#id49" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">ctdt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">ctdt</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span>
<span class="go">array([[0.        , 0.        , 0.25      , 0.5       , 0.        ,</span>
<span class="go">        0.        , 0.25      , 0.5       , 0.        , 0.        ,</span>
<span class="go">        0.5       , 0.        , 0.25      , 0.25      , 0.        ,</span>
<span class="go">        0.25      , 0.25      , 0.25      , 0.25      , 0.        ,</span>
<span class="go">        0.25      , 0.        , 0.25      , 0.5       , 0.        ,</span>
<span class="go">        0.25      , 0.25      , 0.        , 0.25      , 0.5       ,</span>
<span class="go">        0.5       , 0.        , 0.        , 0.25      , 0.        ,</span>
<span class="go">        0.        , 0.5       , 0.25      , 0.        ],</span>
<span class="go">       [0.16666667, 0.33333333, 0.16666667, 0.16666667, 0.        ,</span>
<span class="go">        0.        , 0.16666667, 0.        , 0.16666667, 0.16666667,</span>
<span class="go">        0.        , 0.16666667, 0.5       , 0.        , 0.16666667,</span>
<span class="go">        0.16666667, 0.        , 0.16666667, 0.16666667, 0.33333333,</span>
<span class="go">        0.16666667, 0.16666667, 0.16666667, 0.16666667, 0.16666667,</span>
<span class="go">        0.        , 0.16666667, 0.33333333, 0.        , 0.33333333,</span>
<span class="go">        0.16666667, 0.16666667, 0.        , 0.        , 0.33333333,</span>
<span class="go">        0.        , 0.        , 0.16666667, 0.16666667]])</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span>
<span class="go">[&#39;Hydrophobicity_ARGP820101-T1221&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ARGP820101-T1331&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ARGP820101-T2332&#39;,</span>
<span class="go"> &#39;Hydrophobicity_CASG920101-T1221&#39;,</span>
<span class="go"> &#39;Hydrophobicity_CASG920101-T1331&#39;,</span>
<span class="go"> &#39;Hydrophobicity_CASG920101-T2332&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ENGD860101-T1221&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ENGD860101-T1331&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ENGD860101-T2332&#39;,</span>
<span class="go"> &#39;Hydrophobicity_FASG890101-T1221&#39;,</span>
<span class="go"> &#39;Hydrophobicity_FASG890101-T1331&#39;,</span>
<span class="go"> &#39;Hydrophobicity_FASG890101-T2332&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PONP930101-T1221&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PONP930101-T1331&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PONP930101-T2332&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PRAM900101-T1221&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PRAM900101-T1331&#39;,</span>
<span class="go"> &#39;Hydrophobicity_PRAM900101-T2332&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ZIMJ680101-T1221&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ZIMJ680101-T1331&#39;,</span>
<span class="go"> &#39;Hydrophobicity_ZIMJ680101-T2332&#39;,</span>
<span class="go"> &#39;Normalized van der Waals Volume-T1221&#39;,</span>
<span class="go"> &#39;Normalized van der Waals Volume-T1331&#39;,</span>
<span class="go"> &#39;Normalized van der Waals Volume-T2332&#39;,</span>
<span class="go"> &#39;Polarity-T1221&#39;,</span>
<span class="go"> &#39;Polarity-T1331&#39;,</span>
<span class="go"> &#39;Polarity-T2332&#39;,</span>
<span class="go"> &#39;Polarizability-T1221&#39;,</span>
<span class="go"> &#39;Polarizability-T1331&#39;,</span>
<span class="go"> &#39;Polarizability-T2332&#39;,</span>
<span class="go"> &#39;Charge-T1221&#39;,</span>
<span class="go"> &#39;Charge-T1331&#39;,</span>
<span class="go"> &#39;Charge-T2332&#39;,</span>
<span class="go"> &#39;Secondary structure-T1221&#39;,</span>
<span class="go"> &#39;Secondary structure-T1331&#39;,</span>
<span class="go"> &#39;Secondary structure-T2332&#39;,</span>
<span class="go"> &#39;Solvent accessibility-T1221&#39;,</span>
<span class="go"> &#39;Solvent accessibility-T1331&#39;,</span>
<span class="go"> &#39;Solvent accessibility-T2332&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ctdd">
<h2>ctdd<a class="headerlink" href="#ctdd" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.ctdd(X, *, start=1, end=None)
</pre></div>
</div>
<p>Composition/Transition/Distribution - Distribution</p>
<p>Amino acids are categorized into 3 groups based on their physicochemical
properties. The properties used here include hydrophobicity, normalized van
der Waals volume, polarity, polarizability, charge, secondary structure, and
solvent accessibility. For hydrophobicity, we use seven different groupings
based on different studies, which can all be found in AAIndex1.</p>
<p>There are five distribution descriptors for each physicochemical property
and they are the position percents in the whole sequence for the first
residue, 25% residues, 50% residues, 75% residues, and 100% residues for a
certain encoded class. For instance, if the encoded sequence is
‘32132223311311222222’, then there are 10 residues encoded as 2. The
positions for the first residue 2, the 2nd residue 2 (25% * 10 = 2), the 5th
2 residue (50% * 10 = 5), the 7th 2 (75% * 10 = 7) and the 10th residue 2
(100% * 10) in the encoded sequence are 2, 5, 15, 17, 20, so that the
distribution descriptors for 2 are: 10.0 (2/20 * 100), 25.0 (5/20 * 100),
75.0 (15/20 * 100), 85.0 (17/20 * 100), 100.0 (20/20 * 100).</p>
<p>Since there are 13 physicochemical properties, 3 groups (see ctdc), and 5
distribution descriptors, the total dimensionality of this feature is 195.</p>
<div class="section" id="id50">
<h3>Parameters<a class="headerlink" href="#id50" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id51">
<h3>Returns<a class="headerlink" href="#id51" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, 195)</dt><dd><p>Array containing grouped distribution of physicochemical properties.</p>
</dd>
<dt>desc: list of length 195</dt><dd><p>Order of distribution groups corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id52">
<h3>References<a class="headerlink" href="#id52" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Dubchak, I., Muchnik, I., Holbrook, S. R. &amp; Kim, S.-H. Prediction of protein folding class using global description of amino acid sequence. Proceedings of the National Academy of Sciences 92, 8700–8704 (1995).</p></li>
<li><p>Dubchak, I., Muchnik, I., Mayor, C., Dralyuk, I. &amp; Kim, S.-H. Recognition of a protein fold in the context of the scop classification. Proteins: Structure, Function, and Bioinformatics 35, 401–407 (1999).</p></li>
</ul>
</div>
<div class="section" id="id53">
<h3>Examples<a class="headerlink" href="#id53" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">ctdc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">ctdd</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 195)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">desc</span><span class="p">)</span>
<span class="go">195</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="geary">
<h2>geary<a class="headerlink" href="#geary" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.geary(X, *, d=1, properties=default, start=1, end=None)
</pre></div>
</div>
<p>Geary’s C based on AAIndex1.</p>
<p>Geary’s C autocorrelation descriptors are defined based on the distribution
of AAIndex1-based amino acid properties along the sequence. All indices are
standardized before computing the descriptors. For the exact formula, please
refer to the documentation at <a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>.</p>
<div class="section" id="id54">
<h3>Parameters<a class="headerlink" href="#id54" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>properties: list</dt><dd><p>List of strings denoting AAIndex1 indices.</p>
</dd>
<dt>d: int, default=1</dt><dd><p>Represents the lag. Must be smaller than sequence length. Maximum: 30.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id55">
<h3>Returns<a class="headerlink" href="#id55" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, n_properties)</dt><dd><p>Array containing Geary’s C autocorrelation descriptors.</p>
</dd>
</dl>
</div>
<div class="section" id="id56">
<h3>References<a class="headerlink" href="#id56" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Geary, R. (1954). The Contiguity Ratio and Statistical Mapping. The Incorporated Statistician, 5(3), 115-146. doi:10.2307/2986645</p></li>
<li><p>Jeffers, J. (1973). A Basic Subroutine for Geary’s Contiguity Ratio. Journal of the Royal Statistical Society. Series D (The Statistician), 22(4), 299-302. doi:10.2307/2986827</p></li>
<li><p>Sokal, RR., Thomson, BA. (2006). Population structure inferred by local spatial autocorrelation: an example from an Amerindian tribal population. Am J Phys Anthropol, 129: 121–131. 10.1002/ajpa.20250</p></li>
<li><p>Xiao et al. (2015). protr/ProtrWeb: R package and web server for generating various numerical representation schemes of protein sequences. Bioinformatics 31 (11), 1857-1859.</p></li>
</ul>
</div>
<div class="section" id="id57">
<h3>Examples<a class="headerlink" href="#id57" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">geary</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gearyC</span> <span class="o">=</span> <span class="n">geary</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gearyC</span>
<span class="go">array([[0.52746275, 1.12898944, 0.94222955, 0.39077186, 0.96444569,</span>
<span class="go">        0.66346012, 0.87481962, 0.32546227],</span>
<span class="go">       [0.65656058, 0.95397893, 0.87962853, 0.70972353, 0.65407555,</span>
<span class="go">        0.96823847, 1.01949384, 0.21073089]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="moran">
<h2>moran<a class="headerlink" href="#moran" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.moran(X, *, d=1, properties=default, start=1, end=None)
</pre></div>
</div>
<p>Moran’s I based on AAIndex1.</p>
<p>Moran’s I autocorrelation descriptors are defined based on the distribution
of AAIndex1-based amino acid properties along the sequence. All indices are
standardized before computing the descriptors. For the exact formula, please
refer to the documentation at <a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>.</p>
<div class="section" id="id58">
<h3>Parameters<a class="headerlink" href="#id58" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>properties: list</dt><dd><p>List of strings denoting AAIndex1 indices.</p>
</dd>
<dt>d: int, default=1</dt><dd><p>Represents the lag. Must be smaller than sequence length. Maximum: 30.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id59">
<h3>Returns<a class="headerlink" href="#id59" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, n_properties)</dt><dd><p>Array containing Moran’s I autocorrelation descriptors.</p>
</dd>
</dl>
</div>
<div class="section" id="id60">
<h3>References<a class="headerlink" href="#id60" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Moran, P. (1950). Notes on Continuous Stochastic Phenomena. Biometrika, 37(1/2), 17-23. doi:10.2307/2332142</p></li>
<li><p>Horne, DS. (1988): Prediction of protein helix content from an autocorrelation analysis of sequence hydrophobicities. Biopolymers 27, 451–477</p></li>
<li><p>Li et al. (2007). Beyond Moran’s I: testing for spatial dependence based on the spatial autoregressive model. Geogr. Anal. 39, 357–375.</p></li>
<li><p>Xiao et al. (2015). protr/ProtrWeb: R package and web server for generating various numerical representation schemes of protein sequences. Bioinformatics 31 (11), 1857-1859</p></li>
</ul>
</div>
<div class="section" id="id61">
<h3>Examples<a class="headerlink" href="#id61" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">moran</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moranI</span> <span class="o">=</span> <span class="n">moran</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">moranI</span>
<span class="go">array([[ 0.40090094, -0.31240708, -0.44083728,  0.26720303, -0.45198768,</span>
<span class="go">        -0.14684112, -0.05212843,  0.33703981],</span>
<span class="go">        [-0.0588976 , -0.36033526,  0.13170834,  0.18317369,  0.3884609 ,</span>
<span class="go">         -0.00724234, -0.19231646,  0.61711506]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="moreau-broto">
<h2>moreau_broto<a class="headerlink" href="#moreau-broto" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.moreau_broto(X, *, d=1, properties=default, start=1, end=None)
</pre></div>
</div>
<p>Normalized Moreau-Broto autocorrelation based on AAIndex1.</p>
<p>Moreau-Broto autocorrelation descriptors are defined based on the
distribution of AAIndex1-based amino acid properties along the sequence. All
indices are standardized before computing the descriptors. For the exact
formula, please refer to the documentation at <a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>.</p>
<div class="section" id="id62">
<h3>Parameters<a class="headerlink" href="#id62" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>properties: list</dt><dd><p>List of strings denoting AAIndex1 indices.</p>
</dd>
<dt>d: int, default=1</dt><dd><p>Represents the lag. Must be smaller than sequence length. Maximum: 30.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id63">
<h3>Returns<a class="headerlink" href="#id63" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, n_properties)</dt><dd><p>Array containing Moreau-Broto autocorrelation descriptors.</p>
</dd>
</dl>
</div>
<div class="section" id="id64">
<h3>References<a class="headerlink" href="#id64" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Moreau &amp; Broto (1980), Autocorrelation of a topological structure: A new molecular descriptor, Nouv. J. Chim. 4, 359–360.</p></li>
<li><p>Feng, ZP., Zhang, CT. (2000): Prediction of membrane protein types based on the hydrophobic index of amino acids. J Protein Chem, 19: 262–275. 10.1023/A:1007091128394</p></li>
<li><p>Lin, Z., Pan, XM. (2001): Accurate prediction of protein secondary structural content. J Protein Chem, 20: 217–220. 10.1023/A:1010967008838</p></li>
<li><p>Xiao et al. (2015). protr/ProtrWeb: R package and web server for generating various numerical representation schemes of protein sequences. Bioinformatics 31 (11), 1857-1859</p></li>
</ul>
</div>
<div class="section" id="id65">
<h3>Examples<a class="headerlink" href="#id65" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">moreau_broto</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mb</span> <span class="o">=</span> <span class="n">moreau_broto</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mb</span>
<span class="go">array([[ 0.20075105, -0.19610635, -0.22859247,  0.40627137, -0.19552707,</span>
<span class="go">         0.16803808,  0.11262517,  0.3817956 ],</span>
<span class="go">       [ 0.34693551,  0.55361789,  0.12660283,  0.37790326,  0.43905717,</span>
<span class="go">         0.00559139, -0.09331898,  0.36367721]])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="paac">
<h2>paac<a class="headerlink" href="#paac" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.paac(X, *, lambda_=30, w=.05, start=1, end=None)
</pre></div>
</div>
<p>Pseudo amino acid composition.</p>
<p>Similar to the vanilla amino acid composition, this feature characterizes
the protein mainly using a matrix of amino-acid frequencies, which helps
with dealing with proteins without significant sequence homology to other
proteins. However, additional information are also included in the
matrix to represent some local features, such as correlation between
residues of a certain distance. For the exact method of calculating the
pseudo amino acid composition, please refer to the documentation at
<a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>.</p>
<div class="section" id="id66">
<h3>Parameters<a class="headerlink" href="#id66" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt><a href="#id87"><span class="problematic" id="id88">lambda_</span></a>: int, default=30</dt><dd><p>Counted rank (tier) of the correlation along an amino acid sequence. <br />
This parameter has to be smaller than the shortest sequence in the
dataset.</p>
</dd>
<dt>w: float, default=.05</dt><dd><p>Weighting factor for the sequence-order effect.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id67">
<h3>Returns<a class="headerlink" href="#id67" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr:  ndarray of shape (n_samples, <a href="#id89"><span class="problematic" id="id90">20+lambda_</span></a>)</dt><dd><p>Array containing pseudo amino acid composition.</p>
</dd>
<dt>desc: list of length <a href="#id91"><span class="problematic" id="id92">20+lambda_</span></a></dt><dd><p>Order of amino acids and lambda values corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id68">
<h3>Notes<a class="headerlink" href="#id68" title="Permalink to this headline">¶</a></h3>
<p>Columns containing only zeros will be deleted. Therefore, the returned
array and list may have a lower dimensionality than <a href="#id93"><span class="problematic" id="id94">20+lambda_</span></a>.</p>
</div>
<div class="section" id="id69">
<h3>References<a class="headerlink" href="#id69" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Tanford C. Contribution of hydrophobic interactions to the stability of the globular conformation of proteins. J Am Chem Soc. 1962;84:4240–4247.</p></li>
<li><p>Hopp TP, Woods KR. Prediction of protein antigenic determinants from amino acid sequences. Proc Natl Acad Sci U S A. 1981 Jun;78(6):3824-8. doi: 10.1073/pnas.78.6.3824. PMID: 6167991; PMCID: PMC319665.</p></li>
<li><p>Chou KC. Prediction of protein cellular attributes using pseudo‐amino acid composition. Proteins. 2001; 43: 246‐ 255.</p></li>
<li><p>Jain et al. TpPred: A Tool for Hierarchical Prediction of Transport Proteins Using Cluster of Neural Networks and Sequence Derived Features. International Journal for Computational Biology (IJCB), 2012;1:28-36.</p></li>
</ul>
</div>
<div class="section" id="id70">
<h3>Examples<a class="headerlink" href="#id70" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">paac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">paac_comp</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">paac</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">lambda_</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">paac_comp</span>
<span class="go">array([[0.62956037, 0.        , 0.        , 0.62956037, 0.62956037,</span>
<span class="go">        0.        , 0.62956037, 0.62956037, 0.10830885, 0.16327632,</span>
<span class="go">        0.09885446],</span>
<span class="go">       [0.        , 1.49423911, 0.74711955, 0.74711955, 0.74711955,</span>
<span class="go">        0.74711955, 0.74711955, 0.        , 0.04009348, 0.08251201,</span>
<span class="go">        0.13027496]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span>
<span class="go">[&#39;A&#39;, &#39;E&#39;, &#39;G&#39;, &#39;K&#39;, &#39;L&#39;, &#39;P&#39;, &#39;R&#39;, &#39;Y&#39;, &#39;lambda1&#39;, &#39;lambda2&#39;, &#39;lambda3&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="apaac">
<h2>apaac<a class="headerlink" href="#apaac" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.apaac(X, *, lambda_=30, w=.05, start=1, end=None)
</pre></div>
</div>
<p>Amphiphilic pseudo amino acid composition.</p>
<p>This feature has the same form as the vanilla amino acid composition, but
contains much more information that is related to the sequence order of a
protein and the distribution of the hydrophobic and hydrophilic amino acids
along its chain. For the exact method of calculating the amphiphilic
pseudo amino acid composition, please refer to the documentation at
<a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>.</p>
<div class="section" id="id71">
<h3>Parameters<a class="headerlink" href="#id71" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt><a href="#id95"><span class="problematic" id="id96">lambda_</span></a>: int, default=30</dt><dd><p>Counted rank (tier) of the correlation along an amino acid sequence. <br />
This parameter has to be smaller than the shortest sequence in the
dataset.</p>
</dd>
<dt>w: float, default=.05</dt><dd><p>Weighting factor for the sequence-order effect.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id72">
<h3>Returns<a class="headerlink" href="#id72" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr: ndarray of shape (n_samples, 20+2*lambda_)</dt><dd><p>Array containing amphiphilic pseudo amino acid composition.</p>
</dd>
<dt>desc: list of length 20+2*lambda_</dt><dd><p>Order of amino acids and lambda values corresponding to columns in arr.</p>
</dd>
</dl>
</div>
<div class="section" id="id73">
<h3>Notes<a class="headerlink" href="#id73" title="Permalink to this headline">¶</a></h3>
<p>Columns containing only zeros will be deleted. Therefore, the returned
array and list may have a lower dimensionality than 20+2*lambda_.</p>
</div>
<div class="section" id="id74">
<h3>References<a class="headerlink" href="#id74" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Tanford C. Contribution of hydrophobic interactions to the stability of the globular conformation of proteins. J Am Chem Soc. 1962;84:4240–4247.</p></li>
<li><p>Hopp TP, Woods KR. Prediction of protein antigenic determinants from amino acid sequences. Proc Natl Acad Sci U S A. 1981 Jun;78(6):3824-8. doi: 10.1073/pnas.78.6.3824. PMID: 6167991; PMCID: PMC319665.</p></li>
<li><p>Chou KC. Using amphiphilic pseudo amino acid composition to predict enzyme subfamily classes. Bioinformatics. 2005 Jan 1;21(1):10-9. doi: 10.1093/bioinformatics/bth466. Epub 2004 Aug 12. PMID: 15308540.</p></li>
<li><p>Jain et al. TpPred: A Tool for Hierarchical Prediction of Transport Proteins Using Cluster of Neural Networks and Sequence Derived Features. International Journal for Computational Biology (IJCB), 2012;1:28-36.</p></li>
</ul>
</div>
</div>
<div class="section" id="id75">
<h2>Examples<a class="headerlink" href="#id75" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">apaac</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">apaac_comp</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">apaac</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">lambda_</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">apaac_comp</span>
<span class="go">array([[ 1.15705278e+00,  0.00000000e+00,  0.00000000e+00,</span>
<span class="go">         1.15705278e+00,  1.15705278e+00,  0.00000000e+00,</span>
<span class="go">         1.15705278e+00,  1.15705278e+00,  1.38909208e-02,</span>
<span class="go">         3.13346727e-02, -8.12700103e-02, -6.96352986e-02,</span>
<span class="go">        -1.82775274e-05, -5.13547885e-02],</span>
<span class="go">       [ 0.00000000e+00,  1.64484218e+00,  8.22421091e-01,</span>
<span class="go">         8.22421091e-01,  8.22421091e-01,  8.22421091e-01,</span>
<span class="go">         8.22421091e-01,  0.00000000e+00,  4.76440749e-02,</span>
<span class="go">         6.11246926e-02,  1.80650812e-02,  5.07150233e-02,</span>
<span class="go">        -1.93053341e-02,  1.93353709e-02]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span>
<span class="go">[&#39;A&#39;,</span>
<span class="go"> &#39;E&#39;,</span>
<span class="go"> &#39;G&#39;,</span>
<span class="go"> &#39;K&#39;,</span>
<span class="go"> &#39;L&#39;,</span>
<span class="go"> &#39;P&#39;,</span>
<span class="go"> &#39;R&#39;,</span>
<span class="go"> &#39;Y&#39;,</span>
<span class="go"> &#39;lambda_hphob1&#39;,</span>
<span class="go"> &#39;lambda_hphil1&#39;,</span>
<span class="go"> &#39;lambda_hphob2&#39;,</span>
<span class="go"> &#39;lambda_hphil2&#39;,</span>
<span class="go"> &#39;lambda_hphob3&#39;,</span>
<span class="go"> &#39;lambda_hphil3&#39;]</span>
</pre></div>
</div>
</div>
<div class="section" id="socn">
<h2>socn<a class="headerlink" href="#socn" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.socn(X, *, d=30, start=1, end=None)
</pre></div>
</div>
<p>Sequence-order-coupling number.</p>
<p>This feature is derived from the distance matrix between the 20 amino acids.
Here, we use two different distance matrices based on studies by Grantham
and Schneider-Wrede, respectively. For the exact method of calculating the
sequence-order-coupling number, please refer to the documentation at
<a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>.</p>
<div class="section" id="id76">
<h3>Parameters<a class="headerlink" href="#id76" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>d: int, default=30</dt><dd><p>Represents the lag. Must be smaller than sequence length.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id77">
<h3>Returns<a class="headerlink" href="#id77" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr_sw:  ndarray of shape (n_samples, d)</dt><dd><p>Array containing SOCN based on the Schneider-Wrede distance matrix.</p>
</dd>
<dt>arr_g:  ndarray of shape (n_samples, d)</dt><dd><p>Array containing SOCN based on the Grantham distance matrix.</p>
</dd>
</dl>
</div>
<div class="section" id="id78">
<h3>References<a class="headerlink" href="#id78" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Grantham, R. (1974). Amino acid difference formula to help explain protein evolution. Science. 185 (4154): 862–864</p></li>
<li><p>Schneider &amp; Wrede (1994). The Rational Design of Amino Acid Sequences by Artifical Neural Networks and Simulated Molecular Evolution: De Novo Design of an Idealized Leader Cleavge Site. Biophys Journal, 66, 335-344</p></li>
<li><p>Chou,K.-C. (2000) Prediction of protein subcellar locations by incorporating quasi-sequence-order effect. Biochemical and Biophysical Research Communications, 278, 477–483</p></li>
</ul>
</div>
<div class="section" id="id79">
<h3>Examples<a class="headerlink" href="#id79" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">socn</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sw</span><span class="p">,</span> <span class="n">g</span> <span class="o">=</span> <span class="n">socn</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sw</span>
<span class="go">array([1.138343, 1.64627 , 2.150641])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span>
<span class="go">array([35009., 42394., 38859.])</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="qso">
<h2>qso<a class="headerlink" href="#qso" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>protlearn.features.qso(X, *, d=30, start=1, end=None)
</pre></div>
</div>
<p>Quasi-sequence-order.</p>
<p>This feature is derived from the distance matrix between the 20 amino acids.
Here, we use two different distance matrices based on studies by Grantham
and Schneider-Wrede, respectively. For the exact method of calculating the
quasi-sequence order, please refer to the documentation at
<a class="reference external" href="https://protlearn.readthedocs.io/">https://protlearn.readthedocs.io/</a>.</p>
<div class="section" id="id80">
<h3>Parameters<a class="headerlink" href="#id80" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>X: string, fasta, or a list thereof</dt><dd><p>Dataset of amino acid sequences.</p>
</dd>
<dt>d: int, default=30</dt><dd><p>Represents the lag. Must be smaller than sequence length.</p>
</dd>
<dt>w: float, default=.1</dt><dd><p>Weighting factor for the sequence-order effect.</p>
</dd>
<dt>start: int, default=1</dt><dd><p>Determines the starting point of the amino acid sequence. This number is
based on one-based indexing.</p>
</dd>
<dt>end: int, default=None</dt><dd><p>Determines the end point of the amino acid sequence. Similarly to start,
this number is based on one-based indexing.</p>
</dd>
</dl>
</div>
<div class="section" id="id81">
<h3>Returns<a class="headerlink" href="#id81" title="Permalink to this headline">¶</a></h3>
<dl class="simple">
<dt>arr_sw:  ndarray of shape (n_samples, 20+d)</dt><dd><p>Array containing QSO based on the Schneider-Wrede distance matrix.</p>
</dd>
<dt>arr_g:  ndarray of shape (n_samples, 20+d)</dt><dd><p>Array containing QSO based on the Grantham distance matrix.</p>
</dd>
<dt>desc: list of length 20+d</dt><dd><p>Order of QSO descriptors corresponding to columns in arr_sw and arr_g.</p>
</dd>
</dl>
</div>
<div class="section" id="id82">
<h3>Notes<a class="headerlink" href="#id82" title="Permalink to this headline">¶</a></h3>
<p>Columns containing only zeros will be deleted. Therefore, the returned
arrays and list may have a lower dimensionality than 20+d.</p>
</div>
<div class="section" id="id83">
<h3>References<a class="headerlink" href="#id83" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Grantham, R. (1974). Amino acid difference formula to help explain protein evolution. Science. 185 (4154): 862–864</p></li>
<li><p>Schneider &amp; Wrede (1994). The Rational Design of Amino Acid Sequences by Artifical Neural Networks and Simulated Molecular Evolution: De Novo Design of an Idealized Leader Cleavge Site. Biophys Journal, 66, 335-344</p></li>
<li><p>Chou,K.-C. (2000) Prediction of protein subcellar locations by incorporating quasi-sequence-order effect. Biochemical and Biophysical Research Communications, 278, 477–483</p></li>
</ul>
</div>
<div class="section" id="id84">
<h3>Examples<a class="headerlink" href="#id84" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">protlearn.features</span> <span class="kn">import</span> <span class="n">qso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ARKLY&#39;</span><span class="p">,</span> <span class="s1">&#39;EERKPGL&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sw</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="n">qso</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sw</span>
<span class="go">array([[0.66139001, 0.        , 0.        , 0.66139001, 0.66139001,</span>
<span class="go">        0.        , 0.66139001, 0.66139001, 0.12285782, 0.15604737,</span>
<span class="go">        0.0597048 ],</span>
<span class="go">       [0.        , 1.33911348, 0.66955674, 0.66955674, 0.66955674,</span>
<span class="go">        0.66955674, 0.66955674, 0.        , 0.07621852, 0.11022712,</span>
<span class="go">        0.14399762]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span>
<span class="go">array([[1.42887762e-04, 0.00000000e+00, 0.00000000e+00, 1.42887762e-04,</span>
<span class="go">        1.42887762e-04, 0.00000000e+00, 1.42887762e-04, 1.42887762e-04,</span>
<span class="go">        3.71008073e-01, 4.12445524e-01, 2.16403515e-01],</span>
<span class="go">       [0.00000000e+00, 1.72010458e-04, 8.60052291e-05, 8.60052291e-05,</span>
<span class="go">        8.60052291e-05, 8.60052291e-05, 8.60052291e-05, 0.00000000e+00,</span>
<span class="go">        3.01095707e-01, 3.64610568e-01, 3.34207720e-01]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">desc</span>
<span class="go">[&#39;A&#39;, &#39;E&#39;, &#39;G&#39;, &#39;K&#39;, &#39;L&#39;, &#39;P&#39;, &#39;R&#39;, &#39;Y&#39;, &#39;d1&#39;, &#39;d2&#39;, &#39;d3&#39;]</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="dimensionality_reduction.html" class="btn btn-neutral float-right" title="Dimensionality Reduction &amp; Feature Selection" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="preprocessing.html" class="btn btn-neutral float-left" title="Preprocessing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Thomas Dorfer

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>